src/plots/geo/geo2.js

```diff
@@ -0,0 +1,612 @@
+    // useful to update `geo.center` after interactions
+    var midPt = this.midPt = [
+        b[0][0] + (b[1][0] - b[0][0]) / 2,
+        b[0][1] + (b[1][1] - b[0][1]) / 2
+    ];
+
+    // adjust projection to user setting
+    projection
+        .scale(projLayout.scale * s)
+        .translate([t[0] + (midPt[0] - t[0]), t[1] + (midPt[1] - t[1])])
+        .clipExtent(b);
+
+    // the 'albers usa' projection does not expose a 'center' method
+    // so here's this hack to make it respond to 'geoLayout.center'
+    if(geoLayout._isAlbersUsa) {
+        var centerPx = projection([center.lon, center.lat]);
+        var tt = projection.translate();
+
+        projection.translate([
+            tt[0] - (centerPx[0] - tt[0]),
+            tt[1] - (centerPx[1] - tt[1])
+        ]);
+    }
+};
+
+proto.updateBaseLayers = function(fullLayout, geoLayout) {
+    var _this = this;
+    var topojson = _this.topojson;
+    var layers = _this.layers;
+    var basePaths = _this.basePaths;
+
+    function isAxisLayer(d) {
+        return (d === 'lonaxis' || d === 'lataxis');
+    }
+
+    function isTopoLayer(d) {
+        return (
+            constants.fillLayers.indexOf(d) !== -1 ||
+            constants.lineLayers.indexOf(d) !== -1
+        );
+    }
+
+    var allLayers = this.hasChoropleth ?
+        constants.layersForChoropleth :
+        constants.layers;
+
+    var layerData = allLayers.filter(function(d) {
+        return isTopoLayer(d) ? geoLayout['show' + d] :
+            isAxisLayer(d) ? geoLayout[d].showgrid :
+            true;
+    });
+
+    var join = _this.framework.selectAll('.layer')
+        .data(layerData, String);
+
+    join.exit().each(function(d) {
+        delete layers[d];
+        delete basePaths[d];
+        d3.select(this).remove();
+    });
+
+    join.enter().append('g')
+        .attr('class', function(d) { return 'layer ' + d; })
+        .each(function(d) {
+            var layer = layers[d] = d3.select(this);
+
+            if(d === 'bg') {
+                _this.bgRect = layer.append('rect').style('pointer-events', 'all');
+            } else if(isAxisLayer(d)) {
+                basePaths[d] = layer.append('path');
+            } else if(d === 'backplot') {
+                layer.append('g').classed('choroplethlayer', true);
+            } else if(d === 'frontplot') {
+                layer.append('g').classed('scatterlayer', true);
+            } else if(isTopoLayer(d)) {
+                basePaths[d] = layer.append('path');
+            }
+        });
+
+    join.order();
+
+    join.each(function(d) {
+        var path = basePaths[d];
+        var adj = constants.layerNameToAdjective[d];
+
+        if(d === 'frame') {
+            path.datum(constants.sphereSVG);
+        } else if(isTopoLayer(d)) {
+            path.datum(topojsonFeature(topojson, topojson.objects[d]));
+        } else if(isAxisLayer(d)) {
+            path.datum(makeGraticule(d, geoLayout))
+                .attr('fill', 'none')
```

Comment:
🐎  these constant `.attr` calls can go in `enter()` - but also for consistency switch to `.style` for fill and stroke?

Reply:
done in https://github.com/plotly/plotly.js/pull/2030/commits/eeee9118294af10660a7bcedfbbcea12ec275579